reconcile/terraform_cloudflare_dns.py

```diff
@@ -87,11 +88,43 @@ def run(self, dry_run: bool) -> None:
 
     @defer
     def run_with_defer(self, dry_run: bool, defer: Callable) -> None:
-        vault_settings = get_app_interface_vault_settings()
-        secret_reader = create_secret_reader(use_vault=vault_settings.vault)
+
+        settings = self._get_app_interface_settings()
+
+        if not settings.settings:
+            raise RuntimeError("App interface setting undefined.")
+
+        if not settings.settings[0].vault:
+            raise RuntimeError("App interface vault setting undefined.")
+
+        if not settings.settings[0].cloudflare_dns_zone_max_records:
+            raise RuntimeError(
+                "Default max number of records for Cloudflare DNS zone undefined."
+            )
+
+        default_max_records = settings.settings[0].cloudflare_dns_zone_max_records
+
+        secret_reader = create_secret_reader(use_vault=settings.settings[0].vault)
 
         query_zones = self._get_cloudflare_desired_state()
 
+        if not query_zones.zones:
+            sys.exit(ExitCodes.SUCCESS)
+
+        for zone in query_zones.zones:
+            if not zone.records:
+                continue
+            num_records = len(zone.records)
+            if num_records > default_max_records:
+                if not zone.max_records:
+                    raise RuntimeError(
+                        f"The number of record({num_records}) in zone {zone.identifier} exceeded default max {default_max_records}, please set the max_records and get approval from App SRE."
+                    )
+                if zone.max_records < num_records:
```

Comment:
If we had:

`num_records = 100`
`default_max_records = 1000`
`zone.max_records = 50`

Would this work? Limiting records below the default isn't our primary use case, but in theory it could be used to do so?

Reply:
Updated